File: app/src/main/java/com/yibao/music/fragment/SearchFragment.java
package com.yibao.music.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.yibao.music.R;
import com.yibao.music.activity.PlayListActivity;
import com.yibao.music.adapter.DetailsViewAdapter;
import com.yibao.music.base.BaseFragment;
import com.yibao.music.base.BaseObserver;
import com.yibao.music.base.factory.RecyclerFactory;
import com.yibao.music.base.listener.OnFlowLayoutClickListener;
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog;
import com.yibao.music.model.MoreMenuStatus;
import com.yibao.music.model.MusicBean;
import com.yibao.music.model.SearchCategoryBean;
import com.yibao.music.model.SearchHistoryBean;
import com.yibao.music.util.Constants;
import com.yibao.music.util.LogUtil;
import com.yibao.music.util.MusicDaoUtil;
import com.yibao.music.util.SnakbarUtil;
import com.yibao.music.view.FlowLayoutView;

import java.util.Collections;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

/**
 * @author: Luoshipeng
 * @ Name:    SearchFragment
 * @ Email:   strangermy98@gmail.com
 * @ GitHub:  https://github.com/1900Star
 * @ Time:    2019/3/16/ 16:20
 * @ Des:     TODO
 */
public class SearchFragment extends BaseFragment {
    @BindView(R.id.tv_no_search_result)
    TextView mTvNoSearchResult;
    @BindView(R.id.iv_search_clear)
    ImageView mIvSearchClear;
    @BindView(R.id.sticky_view_search)
    TextView mtvStickyView;
    @BindView(R.id.ll_list_view)
    LinearLayout mLinearDetail;
    @BindView(R.id.ll_history)
    LinearLayout mLayoutHistory;
    @BindView(R.id.flowlayout)
    FlowLayoutView mFlowLayoutView;
    private DetailsViewAdapter mSearchDetailAdapter;
    private Disposable mDisposableSearch;
    private Disposable mDisposableMoreMenu;
    private boolean mIsFirst = false;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.search_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        initData();
        initListener();
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        rxbusData();

    }

    private void rxbusData() {
        if (mDisposableSearch == null) {
            mDisposableSearch = mBus.toObserverable(SearchCategoryBean.class)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(SearchFragment.this::searchSong);
        }
        if (mDisposableMoreMenu == null) {
            mDisposableMoreMenu = mBus.toObserverable(MoreMenuStatus.class)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::moreMenu);
        }
    }

    private void initListener() {
        mSearchDetailAdapter.setOnItemMenuListener((position, musicBean) -> MoreMenuBottomDialog.newInstance(musicBean, position, false, false).getBottomDialog(mActivity));
        mIvSearchClear.setOnClickListener(v -> {
            mSearchDao.deleteAll();
            mFlowLayoutView.removeAllViews();
            mLayoutHistory.setVisibility(View.INVISIBLE);
            mLinearDetail.setVisibility(View.INVISIBLE);
        });
        mFlowLayoutView.setItemClickListener((position, bean) -> initSearch(bean.getSearchContent()));
    }

    /**
     * 点击搜索记录
     *
     * @param searchContent 搜索内容
     */
    private void initSearch(String searchContent) {
        // 填入输入框
        if (mContext instanceof OnFlowLayoutClickListener) {
            ((OnFlowLayoutClickListener) mContext).click(searchContent);
        }
        searchMusic(searchContent);
    }


    private void clearAdapter() {
        if (mSearchDetailAdapter != null) {
            mSearchDetailAdapter.clear();
        }
    }

    /**
     * 搜索音乐
     *
     * @param searchconditions 搜索关键字
     */
    private void searchMusic(String searchconditions) {
        MusicDaoUtil.getSearchResult(mMusicBeanDao, searchconditions)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new BaseObserver<List<MusicBean>>() {
                    @Override
                    public void onNext(List<MusicBean> musicBeanList) {

                        mSearchDetailAdapter.setNewData(musicBeanList);
                        mLayoutHistory.setVisibility(View.GONE);
                        mTvNoSearchResult.setVisibility(View.GONE);
                        mLinearDetail.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void onError(Throwable e) {
                        if (mIsFirst) {
                            mLayoutHistory.setVisibility(View.GONE);
                            mLinearDetail.setVisibility(View.GONE);
                            mTvNoSearchResult.setVisibility(View.VISIBLE);
                        }
                    }

                });
    }

    private void historyViewVisibility() {
        List<SearchHistoryBean> historyList = mSearchDao.queryBuilder().list();
        if (historyList != null && historyList.size() > 0) {
            mLayoutHistory.setVisibility(View.VISIBLE);
            mTvNoSearchResult.setVisibility(View.GONE);
            Collections.sort(historyList);
            mFlowLayoutView.clearView(historyList);
        } else {
            mLayoutHistory.setVisibility(View.GONE);
        }
    }

    private void initData() {
        // 搜索记录
        List<SearchHistoryBean> searchList = mSearchDao.queryBuilder().list();
        if (searchList != null && searchList.size() > 0) {
            mLayoutHistory.setVisibility(View.VISIBLE);
            Collections.sort(searchList);
            mFlowLayoutView.setData(searchList);
        }
        // 列表数据
        mSearchDetailAdapter = new DetailsViewAdapter(mActivity, null, Constants.NUMBER_THRRE);
        RecyclerView recyclerView = RecyclerFactory.creatRecyclerView(1, mSearchDetailAdapter);
        mLinearDetail.addView(recyclerView);
        if (getArguments() != null) {
            int position = getArguments().getInt("position");
            String searchArtist = getArguments().getString("searchArtist");
            setFlagAndSearch(searchArtist, 1);
        }
    }

    public void searchSong(SearchCategoryBean categoryBean) {
        int categoryFlag = categoryBean.getCategoryFlag();
        String condition = categoryBean.getSearchCondition();
        if (condition != null) {
            mIsFirst = true;
        }
        switch (categoryFlag) {
            case 0:
                setFlagAndSearch(condition, Constants.NUMBER_THRRE);
                break;
            case 1:
                setFlagAndSearch(condition, Constants.NUMBER_THRRE);
                break;
            case 2:
                setFlagAndSearch(condition, Constants.NUMBER_TWO);
                break;
            case 3:
                setFlagAndSearch(condition, Constants.NUMBER_ONE);
                break;
            case 4:
                setFlagAndSearch(condition, Constants.NUMBER_FOUR);
                break;
            case 9:
                clearAdapter();
                mLinearDetail.setVisibility(View.INVISIBLE);
                historyViewVisibility();
                break;
            case 10:
                historyViewVisibility();
                clearAdapter();
                mLinearDetail.setVisibility(View.INVISIBLE);
                mTvNoSearchResult.setVisibility(View.GONE);
                break;
            default:
                break;
        }

    }

    private void setFlagAndSearch(String condition, int numberTwo) {
        mSearchDetailAdapter.setDataFlag(numberTwo);
        searchMusic(condition);
    }

    protected void moreMenu(MoreMenuStatus moreMenuStatus) {
        MusicBean musicBean = moreMenuStatus.getMusicBean();
        switch (moreMenuStatus.getPosition()) {
            case Constants.NUMBER_ZERO:
                Intent intent = new Intent(mContext, PlayListActivity.class);
                intent.putExtra(Constants.SONG_NAME, musicBean.getTitle());
                startActivity(intent);
                mActivity.overridePendingTransition(R.anim.dialog_push_in, 0);
                break;
            case Constants.NUMBER_ONE:
                SnakbarUtil.keepGoing(mLayoutHistory);
                break;
            case Constants.NUMBER_TWO:
                break;
            case Constants.NUMBER_THRRE:
                SnakbarUtil.keepGoing(mtvStickyView);
                break;
            case Constants.NUMBER_FOUR:
                // 刷新搜索列表, 后续完成
                mMusicBeanDao.delete(moreMenuStatus.getMusicBean());
                break;
            default:
                break;
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        if (mDisposableSearch != null) {
            mDisposableSearch.dispose();
            mDisposableSearch = null;
        }
        if (mDisposableMoreMenu != null) {
            mDisposableMoreMenu.dispose();
            mDisposableMoreMenu = null;
        }
    }

    public static SearchFragment newInstance(int position, String searchArtist) {
        Bundle args = new Bundle();
        SearchFragment fragment = new SearchFragment();
        args.putInt("position", position);
        args.putString("searchArtist", searchArtist);
        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/com/yibao/music/fragment/SearchFragment.kt
package com.yibao.music.fragment

import android.content.Intent
import android.os.Bundle
import android.view.View
import com.yibao.music.MusicApplication
import com.yibao.music.R
import com.yibao.music.activity.PlayListActivity
import com.yibao.music.adapter.DetailsViewAdapter
import com.yibao.music.base.BaseObserver
import com.yibao.music.base.bindings.BaseBindingAdapter
import com.yibao.music.base.bindings.BaseBindingFragment
import com.yibao.music.base.factory.RecyclerFactory
import com.yibao.music.base.listener.OnFlowLayoutClickListener
import com.yibao.music.databinding.SearchFragmentBinding
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog
import com.yibao.music.model.MoreMenuStatus
import com.yibao.music.model.MusicBean
import com.yibao.music.model.SearchCategoryBean
import com.yibao.music.model.SearchHistoryBean
import com.yibao.music.model.greendao.SearchHistoryBeanDao
import com.yibao.music.util.Constant
import com.yibao.music.util.LogUtil
import com.yibao.music.util.MusicDaoUtil
import com.yibao.music.util.SnakbarUtil
import com.yibao.music.viewmodel.AlbumViewModel
import com.yibao.music.viewmodel.SearchViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import java.util.*

/**
 * @author: Lsp
 * @ Name:    SearchFragment
 * @ GitHub:  https://github.com/1900Star
 * @ Time:    2019/3/16/ 16:20
 * @ Des:     TODO
 */
class SearchFragment : BaseBindingFragment<SearchFragmentBinding>() {
    private lateinit var mSearchDetailAdapter: DetailsViewAdapter
    private var mDisposableSearch: Disposable? = null
    private var mDisposableMoreMenu: Disposable? = null
    private var mIsFirst = false
    private var mPosition = 0
    private lateinit var mSearchDao: SearchHistoryBeanDao
    private val musicList = ArrayList<MusicBean>()

    override fun initView() {
        initData()
        initListener()
        initRecyclerView(mBinding.recyclerSearch)
    }

    override fun initData() {
        mSearchDao = MusicApplication.getInstance().searchDao
        // 搜索记录
        val searchList = mSearchDao.queryBuilder().list()
        if (searchList != null && searchList.size > 0) {
            mBinding.groupSearch.visibility = View.VISIBLE
            searchList.sortBy { it.searchTime }
            mBinding.flowlayout.setData(searchList)
        }
        // 列表数据
        mSearchDetailAdapter = DetailsViewAdapter(mContext, musicList, Constant.NUMBER_THREE)
//        mBinding.recyclerSearch.adapter = mSearchDetailAdapter
        if (arguments != null) {
            mPosition = requireArguments().getInt(Constant.POSITION)
            val searchArtist = requireArguments().getString("searchArtist").toString()
            setFlagAndSearch(searchArtist, 1)
        }
    }


    override fun onResume() {
        super.onResume()
//        rxBusData()

        mViewModel.searchViewModel.observe(this) { bean ->
            if (bean.searchCondition.isNotEmpty()) {
                mPosition = bean.categoryFlag
                searchMusic(bean.searchCondition)
            }
//            searchSong(bean)
        }
    }

    private fun rxBusData() {

        if (mDisposableMoreMenu == null) {
            mDisposableMoreMenu = mBus.toObserverable(MoreMenuStatus::class.java)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { moreMenuStatus: MoreMenuStatus -> moreMenu(moreMenuStatus) }
        }
    }

    private fun initListener() {

        mSearchDetailAdapter.setOnItemMenuListener(object :
            BaseBindingAdapter.OnOpenItemMoreMenuListener {
            override fun openClickMoreMenu(position: Int, musicBean: MusicBean) {
                MoreMenuBottomDialog.newInstance(
                    musicBean,
                    position,
                    false,
                    false
                ).getBottomDialog(requireActivity())

            }
        })

        mBinding.ivSearchClear.setOnClickListener {
            mSearchDao.deleteAll()
            mBinding.flowlayout.removeAllViews()
            mBinding.groupSearch.visibility = View.INVISIBLE
            mBinding.recyclerSearch.visibility = View.INVISIBLE
        }
        mBinding.flowlayout.setItemClickListener { position: Int, bean: SearchHistoryBean ->
            initSearch(
                bean.searchContent
            )
        }
    }

    /**
     * 点击搜索记录
     *
     * @param searchContent 搜索内容
     */
    private fun initSearch(searchContent: String) {
        // 填入输入框
        if (mContext is OnFlowLayoutClickListener) {
            (mContext as OnFlowLayoutClickListener).click(searchContent)
        }
        searchMusic(searchContent)
    }

    private fun clearAdapter() {
        mSearchDetailAdapter.clear()
    }

    /**
     * 搜索音乐
     *
     * @param searchKey 搜索关键字
     */
    private fun searchMusic(searchKey: String) {
        LogUtil.d(mTag, "====key#####################################======= $searchKey")

        if (searchKey.length > 0) {
            MusicDaoUtil.getSearchResult(mMusicBeanDao, searchKey, mPosition)
                .observeOn(AndroidSchedulers.mainThread()).subscribe(object :
                    BaseObserver<List<MusicBean>>() {
                    override fun onNext(list: List<MusicBean>) {
                        list.forEach {
                            when (mPosition) {
                                0 -> {

                                }
                                1 -> {

                                }
                                2 -> {

                                }
                                3 -> {

                                }
                            }
                            LogUtil.d(mTag, it.toString())
                        }
                        musicList.clear()
                        musicList.addAll(list)
                        mSearchDetailAdapter.setNewData(list)
                        mBinding.groupSearch.visibility = View.GONE
                        mBinding.tvNoSearchResult.visibility = View.GONE
                        mBinding.recyclerSearch.visibility = View.VISIBLE

                    }

                    override fun onError(e: Throwable) {
                        if (mIsFirst) {
                            mBinding.groupSearch.visibility = View.GONE
                            mBinding.recyclerSearch.visibility = View.GONE
                            mBinding.tvNoSearchResult.visibility = View.VISIBLE
                        }

                    }
                })
        }


    }

    private fun historyViewVisibility() {
        val historyList = mSearchDao.queryBuilder().list()
        if (historyList != null && historyList.size > 0) {
            mBinding.groupSearch.visibility = View.VISIBLE
            mBinding.tvNoSearchResult.visibility = View.GONE
            with(historyList) { sort() }
            mBinding.flowlayout.clearView(historyList)
        } else {
            mBinding.groupSearch.visibility = View.GONE
        }
    }


    private fun searchSong(categoryBean: SearchCategoryBean) {
        LogUtil.d(mTag, categoryBean.searchCondition)
        val categoryFlag = categoryBean.categoryFlag
        val condition = categoryBean.searchCondition
        if (condition != null) {
            mIsFirst = true
        }
        when (categoryFlag) {
            0 -> setFlagAndSearch(condition, Constant.NUMBER_THREE)
            1 -> setFlagAndSearch(condition, Constant.NUMBER_THREE)
            2 -> setFlagAndSearch(condition, Constant.NUMBER_TWO)
            3 -> setFlagAndSearch(condition, Constant.NUMBER_ONE)
            4 -> setFlagAndSearch(condition, Constant.NUMBER_FOUR)
            9 -> {
                clearAdapter()
                mBinding.recyclerSearch.visibility = View.INVISIBLE
                historyViewVisibility()
            }
            10 -> {
                historyViewVisibility()
                clearAdapter()
                mBinding.recyclerSearch.visibility = View.INVISIBLE
                mBinding.tvNoSearchResult.visibility = View.GONE
            }
            else -> {}
        }
    }

    private fun setFlagAndSearch(condition: String, numberTwo: Int) {
        mSearchDetailAdapter.setDataFlag(numberTwo)
        searchMusic(condition)
    }

    private fun moreMenu(moreMenuStatus: MoreMenuStatus) {
        val musicBean = moreMenuStatus.musicBean
        when (moreMenuStatus.position) {
            Constant.NUMBER_ZERO -> {
                val intent = Intent(mContext, PlayListActivity::class.java)
                intent.putExtra(Constant.SONG_NAME, musicBean.title)
                startActivity(intent)
                requireActivity().overridePendingTransition(R.anim.dialog_push_in, 0)
            }
            Constant.NUMBER_ONE -> SnakbarUtil.keepGoing(mBinding.ivSearchClear)
            Constant.NUMBER_TWO -> {}
            Constant.NUMBER_THREE -> SnakbarUtil.keepGoing(mBinding.ivSearchClear)
            Constant.NUMBER_FOUR ->                 // 刷新搜索列表, 后续完成
                mMusicBeanDao.delete(moreMenuStatus.musicBean)
            else -> {}
        }
    }

    override fun onPause() {
        super.onPause()
        if (mDisposableSearch != null) {
            mDisposableSearch!!.dispose()
            mDisposableSearch = null
        }
        if (mDisposableMoreMenu != null) {
            mDisposableMoreMenu!!.dispose()
            mDisposableMoreMenu = null
        }
    }

    companion object {
        lateinit var mViewModel: SearchViewModel
        fun newInstance(
            position: Int,
            artist: String,
            viewModel: SearchViewModel
        ): SearchFragment {
            mViewModel = viewModel
            val args = Bundle()
            val fragment = SearchFragment()
            args.putInt(Constant.POSITION, position)
            args.putString("searchArtist", artist)
            fragment.arguments = args
            return fragment
        }
    }


}